nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-security/src/main/java/org/apache/nifi/web/security/oidc/StandardOidcIdentityProvider.java

```diff
@@ -0,0 +1,339 @@
+            throw new IllegalStateException(OPEN_ID_CONNECT_SUPPORT_IS_NOT_CONFIGURED);
+        }
+
+        final Scope scope = new Scope("openid");
+
+        // if this provider supports email scope, include it to prevent a subsequent request to the user endpoint
+        if (oidcProviderMetadata.getScopes() != null && oidcProviderMetadata.getScopes().contains(OIDCScopeValue.EMAIL)) {
+            scope.add("email");
+        }
+        return scope;
+    }
+
+    @Override
+    public ClientID getClientId() {
+        if (!isOidcEnabled()) {
+            throw new IllegalStateException(OPEN_ID_CONNECT_SUPPORT_IS_NOT_CONFIGURED);
+        }
+
+        return clientId;
+    }
+
+    @Override
+    public String exchangeAuthorizationCode(final AuthorizationGrant authorizationGrant) throws IOException {
+        if (!isOidcEnabled()) {
+            throw new IllegalStateException(OPEN_ID_CONNECT_SUPPORT_IS_NOT_CONFIGURED);
+        }
+
+        final ClientAuthentication clientAuthentication;
+        if (oidcProviderMetadata.getTokenEndpointAuthMethods().contains(ClientAuthenticationMethod.CLIENT_SECRET_POST)) {
+            clientAuthentication = new ClientSecretPost(clientId, clientSecret);
+        } else {
+            clientAuthentication = new ClientSecretBasic(clientId, clientSecret);
+        }
+
+        try {
+            // build the token request
+            final TokenRequest request = new TokenRequest(oidcProviderMetadata.getTokenEndpointURI(), clientAuthentication, authorizationGrant, getScope());
+            final HTTPRequest tokenHttpRequest = request.toHTTPRequest();
+            tokenHttpRequest.setConnectTimeout(oidcConnectTimeout);
+            tokenHttpRequest.setReadTimeout(oidcReadTimeout);
+
+            // get the token response
+            final TokenResponse response = OIDCTokenResponseParser.parse(tokenHttpRequest.send());
+
+            if (response.indicatesSuccess()) {
+                final OIDCTokenResponse oidcTokenResponse = (OIDCTokenResponse) response;
+                final OIDCTokens oidcTokens = oidcTokenResponse.getOIDCTokens();
+                final JWT oidcJwt = oidcTokens.getIDToken();
+
+                // validate the token - no nonce required for authorization code flow
+                final IDTokenClaimsSet claimsSet = tokenValidator.validate(oidcJwt, null);
+
+                // attempt to extract the email from the id token if possible
+                String email = claimsSet.getStringClaim(EMAIL_CLAIM_NAME);
+                if (email == null) {
+                    // extract the bearer access token
+                    final BearerAccessToken bearerAccessToken = oidcTokens.getBearerAccessToken();
+                    if (bearerAccessToken == null) {
+                        throw new IllegalStateException("No access token found in the ID tokens");
+                    }
+
+                    // invoke the UserInfo endpoint
+                    email = lookupEmail(bearerAccessToken);
```

Comment:
Is there any other unique user identifier that we/users may want to use? If there is an enterprise/internal IdP, is it possible they would provide a `username` or `uid` field but not `email`?